File: src/main/java/dev/tricht/lunaris/util/SystemTray.java
package dev.tricht.lunaris.util;

import dev.tricht.lunaris.Lunaris;
import dev.tricht.lunaris.com.pathofexile.PathOfExileAPI;
import dev.tricht.lunaris.settings.SettingsFrame;
import dev.tricht.lunaris.settings.SettingsWindow;
import javafx.application.Platform;
import lombok.extern.slf4j.Slf4j;
import org.jnativehook.GlobalScreen;
import org.jnativehook.NativeHookException;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ItemEvent;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class SystemTray {

    private static ArrayList<CheckboxMenuItem> leagueMenuItems = new ArrayList<>();
    private static String selectedLeagueName;

    public static String create(PathOfExileAPI pathOfExileAPI) {
        List<String> tradeLeagues = pathOfExileAPI.getTradeLeagues();
        if (!java.awt.SystemTray.isSupported()) {
            if (PropertiesManager.containsKey(PropertiesManager.LEAGUE)) {
                selectedLeagueName = PropertiesManager.getProperty(PropertiesManager.LEAGUE);
            } else {
                selectedLeagueName = tradeLeagues.get(2);
            }
            return selectedLeagueName;
        }
        final PopupMenu popup = new PopupMenu("Lunaris");
        final TrayIcon trayIcon = new TrayIcon(getIcon());
        trayIcon.setImageAutoSize(true);
        final java.awt.SystemTray tray = java.awt.SystemTray.getSystemTray();

        MenuItem POESESSID = new MenuItem("POESESSID");
        POESESSID.addActionListener(e -> {
            String poesessid = (String) JOptionPane.showInputDialog(
                    null,
                    "Enter the POESESSID cookie to prevent rate limits",
                    "Lunaris POESESSID",
                    JOptionPane.PLAIN_MESSAGE,
                    null,
                    null,
                    pathOfExileAPI.getSessionId()
            );
            if (poesessid == null) {
                return;
            }
            pathOfExileAPI.setSessionId(poesessid);
            PropertiesManager.writeProperty(PropertiesManager.POESESSID, poesessid);
        });

        Menu leagueMenu = new Menu("League");
        int count = 0;
        CheckboxMenuItem leagueToSelect = null;
        for (String leagueName : tradeLeagues) {
            CheckboxMenuItem leagueMenuItem = new CheckboxMenuItem(leagueName);
            leagueMenuItems.add(leagueMenuItem);
            leagueMenu.add(leagueMenuItem);
            leagueMenuItem.addItemListener(SystemTray::changeLeagueEventHandler);
            if ((PropertiesManager.containsKey(PropertiesManager.LEAGUE)
                    && PropertiesManager.getProperty(PropertiesManager.LEAGUE).equals(leagueName)) || (count == 2 && leagueToSelect == null)) {
                leagueToSelect = leagueMenuItem;
            }
            count++;
        }
        changeLeagueEventHandler(new ItemEvent(leagueToSelect, 0, leagueToSelect.getLabel(), ItemEvent.SELECTED));

        MenuItem settings = new MenuItem("Settings...");
        settings.addActionListener(e -> {
            SwingUtilities.invokeLater(() -> {
                SettingsWindow window = new SettingsWindow();
                SettingsFrame frame = new SettingsFrame();
                window.add(frame);
                Platform.runLater(() -> {
                    frame.init();
                    window.showSettings(frame.getLayoutBounds());
                });
            });
        });

        MenuItem exitItem = new MenuItem("Exit");

        popup.add(POESESSID);
        popup.add(leagueMenu);
        popup.addSeparator();
        popup.add(settings);
        popup.addSeparator();
        popup.add(exitItem);

        trayIcon.setPopupMenu(popup);

        try {
            tray.add(trayIcon);
        } catch (AWTException e) {
            log.error("TrayIcon could not be added.");
        }

        exitItem.addActionListener(e -> {
            try {
                GlobalScreen.unregisterNativeHook();
            } catch (NativeHookException ex) {
                log.error("Failed to unregister native hook", ex);
            }
            System.exit(0);
        });

        return selectedLeagueName;
    }

    private static Image getIcon() {
        URL imageURL = Lunaris.class.getClassLoader().getResource("icon.png");

        if (imageURL == null) {
            log.error("icon.png missing");
            return null;
        } else {
            return (new ImageIcon(imageURL, "Lunaris")).getImage();
        }
    }

    private static void changeLeagueEventHandler(ItemEvent event) {
        String newLeagueName = event.getItem().toString();
        if (selectedLeagueName != null && selectedLeagueName.equals(newLeagueName)) {
            return;
        }
        selectedLeagueName = event.getItem().toString();
        selectLeague(selectedLeagueName);
        PropertiesManager.writeProperty(PropertiesManager.LEAGUE, selectedLeagueName);
    }

    public static void selectLeague(String league) {
        for (CheckboxMenuItem checkboxMenuItem : leagueMenuItems) {
            checkboxMenuItem.setState(false);
            if (checkboxMenuItem.getLabel().equals(league)) {
                checkboxMenuItem.setState(true);
            }
        }
    }
}

File: src/main/java/dev/tricht/lunaris/util/SystemTray.kt
package dev.tricht.lunaris.util

import dev.tricht.lunaris.com.pathofexile.PathOfExileAPI
import dev.tricht.lunaris.settings.SettingsFrame
import dev.tricht.lunaris.settings.SettingsWindow
import javafx.application.Platform
import org.jnativehook.GlobalScreen
import org.jnativehook.NativeHookException
import org.slf4j.LoggerFactory
import java.awt.*
import java.awt.event.ItemEvent
import java.util.*
import javax.swing.ImageIcon
import javax.swing.SwingUtilities
import kotlin.system.exitProcess

object SystemTray {
    private val leagueMenuItems = ArrayList<CheckboxMenuItem>()
    private val log = LoggerFactory.getLogger(SystemTray::class.java)

    fun setup() {
        if (!java.awt.SystemTray.isSupported()) {
            return;
        }
        val popup = PopupMenu("Lunaris")
        val trayIcon = TrayIcon(icon)
        trayIcon.isImageAutoSize = true
        val tray = java.awt.SystemTray.getSystemTray()
        val leagueMenu = Menu("League")
        var leagueToSelect: CheckboxMenuItem? = null
        for (leagueName in PathOfExileAPI.getTradeLeagues()) {
            val leagueMenuItem = CheckboxMenuItem(leagueName)
            leagueMenuItems.add(leagueMenuItem)
            leagueMenu.add(leagueMenuItem)
            leagueMenuItem.addItemListener(this::changeLeagueEventHandler)
            if (leagueName == Properties.league) {
                leagueToSelect = leagueMenuItem
            }
        }
        changeLeagueEventHandler(ItemEvent(leagueToSelect, 0, leagueToSelect!!.label, ItemEvent.SELECTED))
        val settings = MenuItem("Settings...")
        settings.addActionListener {
            SwingUtilities.invokeLater {
                val window = SettingsWindow()
                val frame = SettingsFrame()
                window.add(frame)
                Platform.runLater {
                    frame.init()
                    window.showSettings(frame.layoutBounds)
                }
            }
        }
        val exitItem = MenuItem("Exit")
        popup.add(leagueMenu)
        popup.addSeparator()
        popup.add(settings)
        popup.addSeparator()
        popup.add(exitItem)
        trayIcon.popupMenu = popup
        try {
            tray.add(trayIcon)
        } catch (e: AWTException) {
            log.error("TrayIcon could not be added.")
        }
        exitItem.addActionListener {
            try {
                GlobalScreen.unregisterNativeHook()
            } catch (ex: NativeHookException) {
                log.error("Failed to unregister native hook", ex)
            }
            exitProcess(0)
        }
    }

    private val icon: Image?
        get() {
            val imageURL = SystemTray::class.java.classLoader.getResource("icon.png")
            return if (imageURL == null) {
                log.error("icon.png missing")
                null
            } else {
                ImageIcon(imageURL, "Lunaris").image
            }
        }

    private fun changeLeagueEventHandler(event: ItemEvent) {
        val newLeagueName = event.item.toString()
        if (Properties.league == newLeagueName) {
            return
        }
        Properties.league = event.item.toString()
        selectLeague(Properties.league)
    }

    fun selectLeague(league: String?) {
        for (checkboxMenuItem in leagueMenuItems) {
            checkboxMenuItem.state = false
            if (checkboxMenuItem.label == league) {
                checkboxMenuItem.state = true
            }
        }
    }
}